libs/Microsoft.MixedReality.WebRTC.Native/src/peer_connection.cpp

```diff
@@ -1188,47 +1463,147 @@ void PeerConnectionImpl::OnAddTrack(
+        global_factory_, *this, transceiver, std::move(audio_track),
+        std::move(receiver), interop_handle);
+    {
+      rtc::CritScope lock(&tracks_mutex_);
+      remote_audio_tracks_.emplace_back(remote_audio_track);
+    }
+
+    // Invoke the AudioTrackAdded callback, which will set the native handle on
+    // the interop wrapper (if created above)
+    {
+      auto lock = std::scoped_lock{media_track_callback_mutex_};
+      auto cb = audio_track_added_callback_;
+      if (cb) {
+        mrsTransceiverHandle tranceiver_handle = transceiver.release();
+        mrsRemoteAudioTrackHandle audio_handle = remote_audio_track.release();
+        cb(interop_handle, audio_handle, transceiver_interop_handle,
+           tranceiver_handle);
+      }
+    }
+  } else if (track_kind_str == webrtc::MediaStreamTrackInterface::kVideoKind) {
+    rtc::scoped_refptr<webrtc::VideoTrackInterface> video_track(
+        static_cast<webrtc::VideoTrackInterface*>(track.release()));
+
+    // Create an interop wrapper for the new native object if needed
+    mrsRemoteVideoTrackInteropHandle interop_handle{};
+    if (auto create_cb = interop_callbacks_.remote_video_track_create_object) {
+      mrsRemoteVideoTrackConfig config;
+      config.track_name = track_name.c_str();
+      interop_handle = (*create_cb)(interop_handle_, config);
+    }
+
+    // Get or create the transceiver wrapper based on the RTP receiver. Because
+    // this callback is fired before the one at the end of the remote
+    // description being applied, the transceiver wrappers for the newly added
+    // RTP transceivers have not been created yet, so create them here.
+    auto ret =
+        GetOrCreateTransceiverForNewRemoteTrack(mrsMediaKind::kVideo, receiver);
+    if (!ret.ok()) {
+      return;
+    }
+    RefPtr<Transceiver> transceiver = ret.MoveValue();
+
+    // The transceiver wrapper might have been created, in which case we need to
+    // inform its interop wrapper of its handle.
+    mrsTransceiverInteropHandle transceiver_interop_handle =
+        transceiver->GetInteropHandle();
+
+    // Create the native object
+    RefPtr<RemoteVideoTrack> remote_video_track = new RemoteVideoTrack(
+        global_factory_, *this, transceiver, std::move(video_track),
+        std::move(receiver), interop_handle);
+    {
+      rtc::CritScope lock(&tracks_mutex_);
+      remote_video_tracks_.emplace_back(remote_video_track);
+    }
+
+    // Invoke the VideoTrackAdded callback, which will set the native handle on
+    // the interop wrapper (if created above)
+    {
+      auto lock = std::scoped_lock{media_track_callback_mutex_};
+      auto cb = video_track_added_callback_;
+      if (cb) {
+        mrsTransceiverHandle tranceiver_handle = transceiver.release();
+        mrsRemoteVideoTrackHandle video_handle = remote_video_track.release();
+        cb(interop_handle, video_handle, transceiver_interop_handle,
+           tranceiver_handle);
+      }
     }
```

Comment:
nit: 
![image](https://user-images.githubusercontent.com/36960428/75702900-4894d600-5cae-11ea-86e0-148ed8baeeb8.png)
This would be nice as two calls to a template function.

Reply:
Same with block in OnRemoveTrack

Reply:
Fixed.